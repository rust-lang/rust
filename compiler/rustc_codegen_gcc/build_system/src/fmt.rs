use std::ffi::OsStr;
use std::path::Path;

use crate::utils::run_command_with_output;

fn show_usage() {
    println!(
        r#"
`fmt` command help:

    --check                : Pass `--check` argument to `cargo fmt` commands
    --help                 : Show this help"#
    );
}

pub fn run() -> Result<(), String> {
    let mut check = false;
    // We skip binary name and the `info` command.
    let mut args = std::env::args().skip(2);
    while let Some(arg) = args.next() {
        match arg.as_str() {
            "--help" => {
                show_usage();
                return Ok(());
            }
            "--check" => check = true,
            _ => return Err(format!("Unknown option {}", arg)),
        }
    }

    let cmd: &[&dyn AsRef<OsStr>] =
        if check { &[&"cargo", &"fmt", &"--check"] } else { &[&"cargo", &"fmt"] };

    run_command_with_output(cmd, Some(&Path::new(".")))?;
    run_command_with_output(cmd, Some(&Path::new("build_system")))
}
