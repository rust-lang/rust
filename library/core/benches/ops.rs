use core::ops::*;

use test::Bencher;

// Overhead of dtors

struct HasDtor {
    _x: isize,
}

impl Drop for HasDtor {
    fn drop(&mut self) {}
}

#[bench]
fn alloc_obj_with_dtor(b: &mut Bencher) {
    b.iter(|| {
        HasDtor { _x: 10 };
    })
}
