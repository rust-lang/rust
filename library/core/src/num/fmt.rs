//! Shared utilities used by both float and integer formatting.
#![doc(hidden)]
#![unstable(
    feature = "numfmt",
    reason = "internal routines only exposed for testing",
    issue = "none"
)]

/// Formatted parts.
#[derive(Copy, Clone, PartialEq, Eq, Debug)]
pub enum Part<'a> {
    /// Given number of zero digits.
    Zero(usize),
    /// A literal number up to 5 digits.
    Num(u16),
    /// A verbatim copy of given bytes.
    Copy(&'a [u8]),
}

impl<'a> Part<'a> {
    /// Returns the exact byte length of given part.
    pub fn len(&self) -> usize {
        match *self {
            Part::Zero(nzeroes) => nzeroes,
            Part::Num(v) => {
                if v < 1_000 {
                    if v < 10 {
                        1
                    } else if v < 100 {
                        2
                    } else {
                        3
                    }
                } else {
                    if v < 10_000 { 4 } else { 5 }
                }
            }
            Part::Copy(buf) => buf.len(),
        }
    }

    /// Writes a part into the supplied buffer.
    /// Returns the number of written bytes, or `None` if the buffer is not enough.
    /// (It may still leave partially written bytes in the buffer; do not rely on that.)
    pub fn write(&self, out: &mut [u8]) -> Option<usize> {
        let len = self.len();
        if out.len() >= len {
            match *self {
                Part::Zero(nzeroes) => {
                    for c in &mut out[..nzeroes] {
                        *c = b'0';
                    }
                }
                Part::Num(mut v) => {
                    for c in out[..len].iter_mut().rev() {
                        *c = b'0' + (v % 10) as u8;
                        v /= 10;
                    }
                }
                Part::Copy(buf) => {
                    out[..buf.len()].copy_from_slice(buf);
                }
            }
            Some(len)
        } else {
            None
        }
    }
}

/// Formatted result containing one or more parts.
/// This can be written to the byte buffer or converted to the allocated string.
#[allow(missing_debug_implementations)]
#[derive(Clone)]
pub struct Formatted<'a> {
    /// A byte slice representing a sign, either `""`, `"-"` or `"+"`.
    pub sign: &'static str,
    /// Formatted parts to be rendered after a sign and optional zero padding.
    pub parts: &'a [Part<'a>],
}

impl<'a> Formatted<'a> {
    /// Returns the exact byte length of combined formatted result.
    pub fn len(&self) -> usize {
        let mut len = self.sign.len();
        for part in self.parts {
            len += part.len();
        }
        len
    }

    /// Writes all formatted parts into the supplied buffer.
    /// Returns the number of written bytes, or `None` if the buffer is not enough.
    /// (It may still leave partially written bytes in the buffer; do not rely on that.)
    pub fn write(&self, out: &mut [u8]) -> Option<usize> {
        if out.len() < self.sign.len() {
            return None;
        }
        out[..self.sign.len()].copy_from_slice(self.sign.as_bytes());

        let mut written = self.sign.len();
        for part in self.parts {
            let len = part.write(&mut out[written..])?;
            written += len;
        }
        Some(written)
    }
}
