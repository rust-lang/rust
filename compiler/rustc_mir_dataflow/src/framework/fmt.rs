//! Custom formatting traits used when outputting Graphviz diagrams with the results of a dataflow
//! analysis.

use std::fmt;

use rustc_index::Idx;
use rustc_index::bit_set::{BitSet, ChunkedBitSet};

use super::lattice::MaybeReachable;

/// An extension to `fmt::Debug` for data that can be better printed with some auxiliary data `C`.
pub trait DebugWithContext<C>: Eq + fmt::Debug {
    fn fmt_with(&self, _ctxt: &C, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }

    /// Print the difference between `self` and `old`.
    ///
    /// This should print nothing if `self == old`.
    ///
    /// `+` and `-` are typically used to indicate differences. However, these characters are
    /// fairly common and may be needed to print a types representation. If using them to indicate
    /// a diff, prefix them with the "Unit Separator"  control character (‚êü  U+001F).
    fn fmt_diff_with(&self, old: &Self, ctxt: &C, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self == old {
            return Ok(());
        }

        write!(f, "\u{001f}+")?;
        self.fmt_with(ctxt, f)?;

        if f.alternate() {
            write!(f, "\n")?;
        } else {
            write!(f, "\t")?;
        }

        write!(f, "\u{001f}-")?;
        old.fmt_with(ctxt, f)
    }
}

/// Implements `fmt::Debug` by deferring to `<T as DebugWithContext<C>>::fmt_with`.
pub struct DebugWithAdapter<'a, T, C> {
    pub this: T,
    pub ctxt: &'a C,
}

impl<T, C> fmt::Debug for DebugWithAdapter<'_, T, C>
where
    T: DebugWithContext<C>,
{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.this.fmt_with(self.ctxt, f)
    }
}

/// Implements `fmt::Debug` by deferring to `<T as DebugWithContext<C>>::fmt_diff_with`.
pub struct DebugDiffWithAdapter<'a, T, C> {
    pub new: T,
    pub old: T,
    pub ctxt: &'a C,
}

impl<T, C> fmt::Debug for DebugDiffWithAdapter<'_, T, C>
where
    T: DebugWithContext<C>,
{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.new.fmt_diff_with(&self.old, self.ctxt, f)
    }
}

// Impls

impl<T, C> DebugWithContext<C> for BitSet<T>
where
    T: Idx + DebugWithContext<C>,
{
    fn fmt_with(&self, ctxt: &C, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_set().entries(self.iter().map(|i| DebugWithAdapter { this: i, ctxt })).finish()
    }

    fn fmt_diff_with(&self, old: &Self, ctxt: &C, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let size = self.domain_size();
        assert_eq!(size, old.domain_size());

        let mut set_in_self = ChunkedBitSet::new_empty(size);
        let mut cleared_in_self = ChunkedBitSet::new_empty(size);

        for i in (0..size).map(T::new) {
            match (self.contains(i), old.contains(i)) {
                (true, false) => set_in_self.insert(i),
                (false, true) => cleared_in_self.insert(i),
                _ => continue,
            };
        }

        fmt_diff(&set_in_self, &cleared_in_self, ctxt, f)
    }
}

impl<T, C> DebugWithContext<C> for ChunkedBitSet<T>
where
    T: Idx + DebugWithContext<C>,
{
    fn fmt_with(&self, ctxt: &C, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_set().entries(self.iter().map(|i| DebugWithAdapter { this: i, ctxt })).finish()
    }

    fn fmt_diff_with(&self, old: &Self, ctxt: &C, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let size = self.domain_size();
        assert_eq!(size, old.domain_size());

        let mut set_in_self = ChunkedBitSet::new_empty(size);
        let mut cleared_in_self = ChunkedBitSet::new_empty(size);

        for i in (0..size).map(T::new) {
            match (self.contains(i), old.contains(i)) {
                (true, false) => set_in_self.insert(i),
                (false, true) => cleared_in_self.insert(i),
                _ => continue,
            };
        }

        fmt_diff(&set_in_self, &cleared_in_self, ctxt, f)
    }
}

impl<S, C> DebugWithContext<C> for MaybeReachable<S>
where
    S: DebugWithContext<C>,
{
    fn fmt_with(&self, ctxt: &C, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            MaybeReachable::Unreachable => {
                write!(f, "unreachable")
            }
            MaybeReachable::Reachable(set) => set.fmt_with(ctxt, f),
        }
    }

    fn fmt_diff_with(&self, old: &Self, ctxt: &C, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match (self, old) {
            (MaybeReachable::Unreachable, MaybeReachable::Unreachable) => Ok(()),
            (MaybeReachable::Unreachable, MaybeReachable::Reachable(set)) => {
                write!(f, "\u{001f}+")?;
                set.fmt_with(ctxt, f)
            }
            (MaybeReachable::Reachable(set), MaybeReachable::Unreachable) => {
                write!(f, "\u{001f}-")?;
                set.fmt_with(ctxt, f)
            }
            (MaybeReachable::Reachable(this), MaybeReachable::Reachable(old)) => {
                this.fmt_diff_with(old, ctxt, f)
            }
        }
    }
}

fn fmt_diff<T, C>(
    inserted: &ChunkedBitSet<T>,
    removed: &ChunkedBitSet<T>,
    ctxt: &C,
    f: &mut fmt::Formatter<'_>,
) -> fmt::Result
where
    T: Idx + DebugWithContext<C>,
{
    let mut first = true;
    for idx in inserted.iter() {
        let delim = if first {
            "\u{001f}+"
        } else if f.alternate() {
            "\n\u{001f}+"
        } else {
            ", "
        };

        write!(f, "{delim}")?;
        idx.fmt_with(ctxt, f)?;
        first = false;
    }

    if !f.alternate() {
        first = true;
        if !inserted.is_empty() && !removed.is_empty() {
            write!(f, "\t")?;
        }
    }

    for idx in removed.iter() {
        let delim = if first {
            "\u{001f}-"
        } else if f.alternate() {
            "\n\u{001f}-"
        } else {
            ", "
        };

        write!(f, "{delim}")?;
        idx.fmt_with(ctxt, f)?;
        first = false;
    }

    Ok(())
}

impl<T, C> DebugWithContext<C> for &'_ T
where
    T: DebugWithContext<C>,
{
    fn fmt_with(&self, ctxt: &C, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        (*self).fmt_with(ctxt, f)
    }

    fn fmt_diff_with(&self, old: &Self, ctxt: &C, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        (*self).fmt_diff_with(*old, ctxt, f)
    }
}

impl<C> DebugWithContext<C> for rustc_middle::mir::Local {}
impl<C> DebugWithContext<C> for crate::move_paths::InitIndex {}

impl<'tcx, C> DebugWithContext<C> for crate::move_paths::MovePathIndex
where
    C: crate::move_paths::HasMoveData<'tcx>,
{
    fn fmt_with(&self, ctxt: &C, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", ctxt.move_data().move_paths[*self])
    }
}
