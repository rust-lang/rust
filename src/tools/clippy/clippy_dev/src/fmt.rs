use crate::utils::clippy_project_root;
use itertools::Itertools;
use rustc_lexer::{TokenKind, tokenize};
use shell_escape::escape;
use std::ffi::{OsStr, OsString};
use std::ops::ControlFlow;
use std::path::{Path, PathBuf};
use std::process::{self, Command, Stdio};
use std::{fs, io};
use walkdir::WalkDir;

pub enum Error {
    CommandFailed(String, String),
    Io(io::Error),
    RustfmtNotInstalled,
    WalkDir(walkdir::Error),
    IntellijSetupActive,
    Parse(PathBuf, usize, String),
    CheckFailed,
}

impl From<io::Error> for Error {
    fn from(error: io::Error) -> Self {
        Self::Io(error)
    }
}

impl From<walkdir::Error> for Error {
    fn from(error: walkdir::Error) -> Self {
        Self::WalkDir(error)
    }
}

impl Error {
    fn display(&self) {
        match self {
            Self::CheckFailed => {
                eprintln!("Formatting check failed!\nRun `cargo dev fmt` to update.");
            },
            Self::CommandFailed(command, stderr) => {
                eprintln!("error: command `{command}` failed!\nstderr: {stderr}");
            },
            Self::Io(err) => {
                eprintln!("error: {err}");
            },
            Self::RustfmtNotInstalled => {
                eprintln!("error: rustfmt nightly is not installed.");
            },
            Self::WalkDir(err) => {
                eprintln!("error: {err}");
            },
            Self::IntellijSetupActive => {
                eprintln!(
                    "error: a local rustc repo is enabled as path dependency via `cargo dev setup intellij`.\n\
                    Not formatting because that would format the local repo as well!\n\
                    Please revert the changes to `Cargo.toml`s with `cargo dev remove intellij`."
                );
            },
            Self::Parse(path, line, msg) => {
                eprintln!("error parsing `{}:{line}`: {msg}", path.display());
            },
        }
    }
}

struct FmtContext {
    check: bool,
    verbose: bool,
    rustfmt_path: String,
}

struct ClippyConf<'a> {
    name: &'a str,
    attrs: &'a str,
    lints: Vec<&'a str>,
    field: &'a str,
}

fn offset_to_line(text: &str, offset: usize) -> usize {
    match text.split('\n').try_fold((1usize, 0usize), |(line, pos), s| {
        let pos = pos + s.len() + 1;
        if pos > offset {
            ControlFlow::Break(line)
        } else {
            ControlFlow::Continue((line + 1, pos))
        }
    }) {
        ControlFlow::Break(x) | ControlFlow::Continue((x, _)) => x,
    }
}

/// Formats the configuration list in `clippy_config/src/conf.rs`
#[expect(clippy::too_many_lines)]
fn fmt_conf(check: bool) -> Result<(), Error> {
    #[derive(Clone, Copy)]
    enum State {
        Start,
        Docs,
        Pound,
        OpenBracket,
        Attr(u32),
        Lints,
        EndLints,
        Field,
    }

    let path: PathBuf = [
        clippy_project_root().as_path(),
        "clippy_config".as_ref(),
        "src".as_ref(),
        "conf.rs".as_ref(),
    ]
    .into_iter()
    .collect();
    let text = fs::read_to_string(&path)?;

    let (pre, conf) = text
        .split_once("define_Conf! {\n")
        .expect("can't find config definition");
    let (conf, post) = conf.split_once("\n}\n").expect("can't find config definition");
    let conf_offset = pre.len() + 15;

    let mut pos = 0u32;
    let mut attrs_start = 0;
    let mut attrs_end = 0;
    let mut field_start = 0;
    let mut lints = Vec::new();
    let mut name = "";
    let mut fields = Vec::new();
    let mut state = State::Start;

    for (i, t) in tokenize(conf)
        .map(|x| {
            let start = pos;
            pos += x.len;
            (start as usize, x)
        })
        .filter(|(_, t)| !matches!(t.kind, TokenKind::Whitespace))
    {
        match (state, t.kind) {
            (State::Start, TokenKind::LineComment { doc_style: Some(_) }) => {
                attrs_start = i;
                attrs_end = i + t.len as usize;
                state = State::Docs;
            },
            (State::Start, TokenKind::Pound) => {
                attrs_start = i;
                attrs_end = i;
                state = State::Pound;
            },
            (State::Docs, TokenKind::LineComment { doc_style: Some(_) }) => attrs_end = i + t.len as usize,
            (State::Docs, TokenKind::Pound) => state = State::Pound,
            (State::Pound, TokenKind::OpenBracket) => state = State::OpenBracket,
            (State::OpenBracket, TokenKind::Ident) => {
                state = if conf[i..i + t.len as usize] == *"lints" {
                    State::Lints
                } else {
                    State::Attr(0)
                };
            },
            (State::Attr(0), TokenKind::CloseBracket) => {
                attrs_end = i + 1;
                state = State::Docs;
            },
            (State::Attr(x), TokenKind::OpenParen | TokenKind::OpenBracket | TokenKind::OpenBrace) => {
                state = State::Attr(x + 1);
            },
            (State::Attr(x), TokenKind::CloseParen | TokenKind::CloseBracket | TokenKind::CloseBrace) => {
                state = State::Attr(x - 1);
            },
            (State::Lints, TokenKind::Ident) => lints.push(&conf[i..i + t.len as usize]),
            (State::Lints, TokenKind::CloseBracket) => state = State::EndLints,
            (State::EndLints | State::Docs, TokenKind::Ident) => {
                field_start = i;
                name = &conf[i..i + t.len as usize];
                state = State::Field;
            },
            (State::Field, TokenKind::LineComment { doc_style: Some(_) }) => {
                #[expect(clippy::drain_collect)]
                fields.push(ClippyConf {
                    name,
                    lints: lints.drain(..).collect(),
                    attrs: &conf[attrs_start..attrs_end],
                    field: conf[field_start..i].trim_end(),
                });
                attrs_start = i;
                attrs_end = i + t.len as usize;
                state = State::Docs;
            },
            (State::Field, TokenKind::Pound) => {
                #[expect(clippy::drain_collect)]
                fields.push(ClippyConf {
                    name,
                    lints: lints.drain(..).collect(),
                    attrs: &conf[attrs_start..attrs_end],
                    field: conf[field_start..i].trim_end(),
                });
                attrs_start = i;
                attrs_end = i;
                state = State::Pound;
            },
            (State::Field | State::Attr(_), _)
            | (State::Lints, TokenKind::Comma | TokenKind::OpenParen | TokenKind::CloseParen) => {},
            _ => {
                return Err(Error::Parse(
                    path,
                    offset_to_line(&text, conf_offset + i),
                    format!("unexpected token `{}`", &conf[i..i + t.len as usize]),
                ));
            },
        }
    }

    if !matches!(state, State::Field) {
        return Err(Error::Parse(
            path,
            offset_to_line(&text, conf_offset + conf.len()),
            "incomplete field".into(),
        ));
    }
    fields.push(ClippyConf {
        name,
        lints,
        attrs: &conf[attrs_start..attrs_end],
        field: conf[field_start..].trim_end(),
    });

    for field in &mut fields {
        field.lints.sort_unstable();
    }
    fields.sort_by_key(|x| x.name);

    let new_text = format!(
        "{pre}define_Conf! {{\n{}}}\n{post}",
        fields.iter().format_with("", |field, f| {
            if field.lints.is_empty() {
                f(&format_args!("    {}\n    {}\n", field.attrs, field.field))
            } else if field.lints.iter().map(|x| x.len() + 2).sum::<usize>() < 120 - 14 {
                f(&format_args!(
                    "    {}\n    #[lints({})]\n    {}\n",
                    field.attrs,
                    field.lints.iter().join(", "),
                    field.field,
                ))
            } else {
                f(&format_args!(
                    "    {}\n    #[lints({}\n    )]\n    {}\n",
                    field.attrs,
                    field
                        .lints
                        .iter()
                        .format_with("", |x, f| f(&format_args!("\n        {x},"))),
                    field.field,
                ))
            }
        })
    );

    if text != new_text {
        if check {
            return Err(Error::CheckFailed);
        }
        fs::write(&path, new_text.as_bytes())?;
    }
    Ok(())
}

fn run_rustfmt(context: &FmtContext) -> Result<(), Error> {
    let project_root = clippy_project_root();

    // if we added a local rustc repo as path dependency to clippy for rust analyzer, we do NOT want to
    // format because rustfmt would also format the entire rustc repo as it is a local
    // dependency
    if fs::read_to_string(project_root.join("Cargo.toml"))
        .expect("Failed to read clippy Cargo.toml")
        .contains("[target.'cfg(NOT_A_PLATFORM)'.dependencies]")
    {
        return Err(Error::IntellijSetupActive);
    }

    check_for_rustfmt(context)?;

    cargo_fmt(context, project_root.as_path())?;
    cargo_fmt(context, &project_root.join("clippy_dev"))?;
    cargo_fmt(context, &project_root.join("rustc_tools_util"))?;
    cargo_fmt(context, &project_root.join("lintcheck"))?;

    let chunks = WalkDir::new(project_root.join("tests"))
        .into_iter()
        .filter_map(|entry| {
            let entry = entry.expect("failed to find tests");
            let path = entry.path();

            if path.extension() != Some("rs".as_ref()) || entry.file_name() == "ice-3891.rs" {
                None
            } else {
                Some(entry.into_path().into_os_string())
            }
        })
        .chunks(250);

    for chunk in &chunks {
        rustfmt(context, chunk)?;
    }
    Ok(())
}

// the "main" function of cargo dev fmt
pub fn run(check: bool, verbose: bool) {
    let output = Command::new("rustup")
        .args(["which", "rustfmt"])
        .stderr(Stdio::inherit())
        .output()
        .expect("error running `rustup which rustfmt`");
    if !output.status.success() {
        eprintln!("`rustup which rustfmt` did not execute successfully");
        process::exit(1);
    }
    let mut rustfmt_path = String::from_utf8(output.stdout).expect("invalid rustfmt path");
    rustfmt_path.truncate(rustfmt_path.trim_end().len());

    let context = FmtContext {
        check,
        verbose,
        rustfmt_path,
    };
    if let Err(e) = run_rustfmt(&context).and_then(|()| fmt_conf(check)) {
        e.display();
        process::exit(1);
    }
}

fn format_command(program: impl AsRef<OsStr>, dir: impl AsRef<Path>, args: &[impl AsRef<OsStr>]) -> String {
    let arg_display: Vec<_> = args.iter().map(|a| escape(a.as_ref().to_string_lossy())).collect();

    format!(
        "cd {} && {} {}",
        escape(dir.as_ref().to_string_lossy()),
        escape(program.as_ref().to_string_lossy()),
        arg_display.join(" ")
    )
}

fn exec_fmt_command(
    context: &FmtContext,
    program: impl AsRef<OsStr>,
    dir: impl AsRef<Path>,
    args: &[impl AsRef<OsStr>],
) -> Result<(), Error> {
    if context.verbose {
        println!("{}", format_command(&program, &dir, args));
    }

    let output = Command::new(&program)
        .env("RUSTFMT", &context.rustfmt_path)
        .current_dir(&dir)
        .args(args.iter())
        .output()
        .unwrap();
    let success = output.status.success();

    match (context.check, success) {
        (_, true) => Ok(()),
        (true, false) => Err(Error::CheckFailed),
        (false, false) => {
            let stderr = std::str::from_utf8(&output.stderr).unwrap_or("");
            Err(Error::CommandFailed(
                format_command(&program, &dir, args),
                String::from(stderr),
            ))
        },
    }
}

fn cargo_fmt(context: &FmtContext, path: &Path) -> Result<(), Error> {
    let mut args = vec!["fmt", "--all"];
    if context.check {
        args.push("--check");
    }
    exec_fmt_command(context, "cargo", path, &args)
}

fn check_for_rustfmt(context: &FmtContext) -> Result<(), Error> {
    let program = "rustfmt";
    let dir = std::env::current_dir()?;
    let args = &["--version"];

    if context.verbose {
        println!("{}", format_command(program, &dir, args));
    }

    let output = Command::new(program).current_dir(&dir).args(args.iter()).output()?;

    if output.status.success() {
        Ok(())
    } else if std::str::from_utf8(&output.stderr)
        .unwrap_or("")
        .starts_with("error: 'rustfmt' is not installed")
    {
        Err(Error::RustfmtNotInstalled)
    } else {
        Err(Error::CommandFailed(
            format_command(program, &dir, args),
            std::str::from_utf8(&output.stderr).unwrap_or("").to_string(),
        ))
    }
}

fn rustfmt(context: &FmtContext, paths: impl Iterator<Item = OsString>) -> Result<(), Error> {
    let mut args = Vec::new();
    if context.check {
        args.push(OsString::from("--check"));
    }
    args.extend(paths);
    exec_fmt_command(context, &context.rustfmt_path, std::env::current_dir()?, &args)
}
