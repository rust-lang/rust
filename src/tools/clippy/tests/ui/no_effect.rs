#![feature(fn_traits, unboxed_closures)]
#![warn(clippy::no_effect_underscore_binding)]
#![allow(
    clippy::deref_addrof,
    clippy::redundant_field_names,
    clippy::uninlined_format_args,
    clippy::unnecessary_struct_initialization,
    clippy::useless_vec
)]

use std::fmt::Display;
use std::ops::{Neg, Shl};

struct Cout;

impl<T> Shl<T> for Cout
where
    T: Display,
{
    type Output = Self;
    fn shl(self, rhs: T) -> Self::Output {
        println!("{}", rhs);
        self
    }
}

impl Neg for Cout {
    type Output = Self;
    fn neg(self) -> Self::Output {
        println!("hello world");
        self
    }
}

struct Tuple(i32);
struct Struct {
    field: i32,
}
enum Enum {
    Tuple(i32),
    Struct { field: i32 },
}
struct DropStruct {
    field: i32,
}
impl Drop for DropStruct {
    fn drop(&mut self) {}
}
struct DropTuple(i32);
impl Drop for DropTuple {
    fn drop(&mut self) {}
}
enum DropEnum {
    Tuple(i32),
    Struct { field: i32 },
}
impl Drop for DropEnum {
    fn drop(&mut self) {}
}
struct FooString {
    s: String,
}
union Union {
    a: u8,
    b: f64,
}

fn get_number() -> i32 {
    0
}
fn get_struct() -> Struct {
    Struct { field: 0 }
}
fn get_drop_struct() -> DropStruct {
    DropStruct { field: 0 }
}

unsafe fn unsafe_fn() -> i32 {
    0
}

struct GreetStruct1;

impl FnOnce<(&str,)> for GreetStruct1 {
    type Output = ();

    extern "rust-call" fn call_once(self, (who,): (&str,)) -> Self::Output {
        println!("hello {}", who);
    }
}

struct GreetStruct2();

impl FnOnce<(&str,)> for GreetStruct2 {
    type Output = ();

    extern "rust-call" fn call_once(self, (who,): (&str,)) -> Self::Output {
        println!("hello {}", who);
    }
}

struct GreetStruct3;

impl FnOnce<(&str,)> for GreetStruct3 {
    type Output = ();

    extern "rust-call" fn call_once(self, (who,): (&str,)) -> Self::Output {
        println!("hello {}", who);
    }
}

fn main() {
    let s = get_struct();

    0;
    //~^ ERROR: statement with no effect
    Tuple(0);
    //~^ ERROR: statement with no effect
    Struct { field: 0 };
    //~^ ERROR: statement with no effect
    Struct { ..s };
    //~^ ERROR: statement with no effect
    Union { a: 0 };
    //~^ ERROR: statement with no effect
    Enum::Tuple(0);
    //~^ ERROR: statement with no effect
    Enum::Struct { field: 0 };
    //~^ ERROR: statement with no effect
    5 + 6;
    //~^ ERROR: statement with no effect
    *&42;
    //~^ ERROR: statement with no effect
    &6;
    //~^ ERROR: statement with no effect
    (5, 6, 7);
    //~^ ERROR: statement with no effect
    ..;
    //~^ ERROR: statement with no effect
    5..;
    //~^ ERROR: statement with no effect
    ..5;
    //~^ ERROR: statement with no effect
    5..6;
    //~^ ERROR: statement with no effect
    5..=6;
    //~^ ERROR: statement with no effect
    [42, 55];
    //~^ ERROR: statement with no effect
    [42, 55][1];
    //~^ ERROR: statement with no effect
    (42, 55).1;
    //~^ ERROR: statement with no effect
    [42; 55];
    //~^ ERROR: statement with no effect
    [42; 55][13];
    //~^ ERROR: statement with no effect
    let mut x = 0;
    || x += 5;
    //~^ ERROR: statement with no effect
    let s: String = "foo".into();
    FooString { s: s };
    //~^ ERROR: statement with no effect
    let _unused = 1;
    //~^ ERROR: binding to `_` prefixed variable with no side-effect
    //~| NOTE: `-D clippy::no-effect-underscore-binding` implied by `-D warnings`
    let _penguin = || println!("Some helpful closure");
    //~^ ERROR: binding to `_` prefixed variable with no side-effect
    let _duck = Struct { field: 0 };
    //~^ ERROR: binding to `_` prefixed variable with no side-effect
    let _cat = [2, 4, 6, 8][2];
    //~^ ERROR: binding to `_` prefixed variable with no side-effect
    let _issue_12166 = 42;
    let underscore_variable_above_can_be_used_dont_lint = _issue_12166;

    #[allow(clippy::no_effect)]
    0;

    // Do not warn
    get_number();
    unsafe { unsafe_fn() };
    let _used = get_struct();
    let _x = vec![1];
    DropStruct { field: 0 };
    DropTuple(0);
    DropEnum::Tuple(0);
    DropEnum::Struct { field: 0 };
    GreetStruct1("world");
    GreetStruct2()("world");
    GreetStruct3 {}("world");

    fn n() -> i32 {
        42
    }

    Cout << 142;
    -Cout;
}
