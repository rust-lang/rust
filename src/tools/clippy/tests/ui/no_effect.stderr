error: statement with no effect
  --> tests/ui/no_effect.rs:115:5
   |
LL |     0;
   |     ^^
   |
   = note: `-D clippy::no-effect` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::no_effect)]`

error: statement with no effect
  --> tests/ui/no_effect.rs:117:5
   |
LL |     Tuple(0);
   |     ^^^^^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:119:5
   |
LL |     Struct { field: 0 };
   |     ^^^^^^^^^^^^^^^^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:121:5
   |
LL |     Struct { ..s };
   |     ^^^^^^^^^^^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:123:5
   |
LL |     Union { a: 0 };
   |     ^^^^^^^^^^^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:125:5
   |
LL |     Enum::Tuple(0);
   |     ^^^^^^^^^^^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:127:5
   |
LL |     Enum::Struct { field: 0 };
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:129:5
   |
LL |     5 + 6;
   |     ^^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:131:5
   |
LL |     *&42;
   |     ^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:133:5
   |
LL |     &6;
   |     ^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:135:5
   |
LL |     (5, 6, 7);
   |     ^^^^^^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:137:5
   |
LL |     ..;
   |     ^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:139:5
   |
LL |     5..;
   |     ^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:141:5
   |
LL |     ..5;
   |     ^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:143:5
   |
LL |     5..6;
   |     ^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:145:5
   |
LL |     5..=6;
   |     ^^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:147:5
   |
LL |     [42, 55];
   |     ^^^^^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:149:5
   |
LL |     [42, 55][1];
   |     ^^^^^^^^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:151:5
   |
LL |     (42, 55).1;
   |     ^^^^^^^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:153:5
   |
LL |     [42; 55];
   |     ^^^^^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:155:5
   |
LL |     [42; 55][13];
   |     ^^^^^^^^^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:158:5
   |
LL |     || x += 5;
   |     ^^^^^^^^^^

error: statement with no effect
  --> tests/ui/no_effect.rs:161:5
   |
LL |     FooString { s: s };
   |     ^^^^^^^^^^^^^^^^^^^

error: binding to `_` prefixed variable with no side-effect
  --> tests/ui/no_effect.rs:163:9
   |
LL |     let _unused = 1;
   |         ^^^^^^^
   |
   = note: `-D clippy::no-effect-underscore-binding` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::no_effect_underscore_binding)]`

error: binding to `_` prefixed variable with no side-effect
  --> tests/ui/no_effect.rs:166:9
   |
LL |     let _penguin = || println!("Some helpful closure");
   |         ^^^^^^^^

error: binding to `_` prefixed variable with no side-effect
  --> tests/ui/no_effect.rs:168:9
   |
LL |     let _duck = Struct { field: 0 };
   |         ^^^^^

error: binding to `_` prefixed variable with no side-effect
  --> tests/ui/no_effect.rs:170:9
   |
LL |     let _cat = [2, 4, 6, 8][2];
   |         ^^^^

error: aborting due to 27 previous errors

